wlc/__init__.py

```diff
@@ -704,7 +704,7 @@ def upload(self, file, overwrite=None, **kwargs):
         """Updoad a translation file to server."""
         self.ensure_loaded("file_url")
         url = self._attribs["file_url"]
-        files = {"file": file}
+        files = {"file": (f"file.{kwargs.get('format','po')}", file)}
```

Comment:
Did you check why the filename is actually needed? It should be only necessary when uploading using other format to make the autodetection possible.

Reply:
This is the traceback we got back from weblate v4.4.2

``` 
weblate_1   | uwsgi stderr | [2021-02-19 11:09:36,703: WARNING/382] Upload error: ValueError: Unknown filetype (file)
weblate_1   | uwsgi stderr | [2021-02-19 11:09:36,703: ERROR/382] Upload error
weblate_1   | uwsgi stderr | Traceback (most recent call last):
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/translate/storage/factory.py", line 191, in getclass
weblate_1   | uwsgi stderr |     storeclass = import_class(*classes_str[ext], "translate.storage")
weblate_1   | uwsgi stderr | KeyError: ''
weblate_1   | uwsgi stderr | 
weblate_1   | uwsgi stderr | During handling of the above exception, another exception occurred:
weblate_1   | uwsgi stderr | 
weblate_1   | uwsgi stderr | Traceback (most recent call last):
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/api/views.py", line 1105, in file
weblate_1   | uwsgi stderr |     data["fuzzy"],
weblate_1   | uwsgi stderr |   File "/usr/lib/python3.7/contextlib.py", line 74, in inner
weblate_1   | uwsgi stderr |     return func(*args, **kwds)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/trans/models/translation.py", line 1010, in merge_upload
weblate_1   | uwsgi stderr |     self.component.template_store,
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/auto.py", line 77, in try_load
weblate_1   | uwsgi stderr |     raise failure
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/auto.py", line 71, in try_load
weblate_1   | uwsgi stderr |     result = file_format.parse(BytesIOMode(filename, content))
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/auto.py", line 103, in parse
weblate_1   | uwsgi stderr |     return cls(storefile, template_store, language_code, is_template)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/ttkit.py", line 220, in __init__
weblate_1   | uwsgi stderr |     super().__init__(storefile, template_store, language_code, is_template)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/base.py", line 197, in __init__
weblate_1   | uwsgi stderr |     self.store = self.load(storefile, template_store)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/ttkit.py", line 244, in load
weblate_1   | uwsgi stderr |     return cls.parse_store(storefile)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/weblate/formats/auto.py", line 108, in parse_store
weblate_1   | uwsgi stderr |     return factory.getobject(storefile)
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/translate/storage/factory.py", line 228, in getobject
weblate_1   | uwsgi stderr |     hiddenclasses=hiddenclasses,
weblate_1   | uwsgi stderr |   File "/usr/local/lib/python3.7/dist-packages/translate/storage/factory.py", line 193, in getclass
weblate_1   | uwsgi stderr |     raise ValueError("Unknown filetype (%s)" % storefilename)
weblate_1   | uwsgi stderr | ValueError: Unknown filetype (file)
``` 

Our call is

```python
        resp = translation.upload(
            new_po.__unicode__(), # Unicode PO content
            overwrite=True,  # This might be deprecated at this point?
            conflicts="replace-translated",  # future proofing for when overwrite is deprecated
            method="translate",
            format="po",
        )

```

Reply:
That looks like something to be addressed in Weblate rather than here. I think that https://github.com/WeblateOrg/weblate/commit/b167285d7eda534d3e40c8675842839445c446a9 should address it for your case.

Anyway if you want to get this in, it should only change behavior when format is present, without that the file should be uploaded without filename to avoid implicit fallback to ``po``.

Reply:
I adapted the code acording to your suggestion of not defaulting to PO but to actually what is passed to the kwargs.

Thanks